pkg/client/informers/informers_generated/internalversion/factory.go

```diff
@@ -98,8 +111,9 @@ func (f *sharedInformerFactory) WaitForCacheSync(stopCh <-chan struct{}) map[ref
 	return res
 }
 
-// InternalInformerFor returns the SharedIndexInformer for obj using an internal
-// client.
+// InformerFor returns the SharedIndexInformer for obj. If the factory has already been started and
+// this is the first time this particular informer has been referenced, the informer will be
+// automatically started. Otherwise, the informer will be started when Start() is invoked.
 func (f *sharedInformerFactory) InformerFor(obj runtime.Object, newFunc internalinterfaces.NewInformerFunc) cache.SharedIndexInformer {
```

Comment:
This is a change of contract which may break people. E.g. it will be impossible to use this method and then add an indexer to the returned informer if the factory has been started already - you cannot add an indexer to a running informer. Just saying.

Reply:
Good point. We need to decide if that's ok. I think I'd rather solve https://github.com/kubernetes/kubernetes/issues/48198 for real, instead of relying on `...Informer().AddIndexers()`.

Reply:
>E.g. it will be impossible to use this method and then add an indexer to the returned informer if the factory has been started already

How would you safely do this today given the caller has no way of knowing whether the factory has started the informer(s)? Are there any existing cases like this in the codebase?

I kind of question whether there really is a strong contract defined in the first place. Here are the extent of the docs:

```
// InternalInformerFor returns the SharedIndexInformer for obj using an internal
// client.
```
And the interface itself actually has no docs at all.

In terms of the concrete impl we're using, the semantics are pretty weird. For example, today, controller A could get a new informer, controller B could get a new informer (neither are started), and then controller A could arbitrarily start both informers at any time via the factory. There's no mechanism for controllers to know whether the factory is started, and they have no choice but to assume start could be called at any point in time asynchronously.

I think it should be possible to control the lifecycle of the informers managed by the factory from outside the controllers, and this PR seems to brings us a step closer to that.

Reply:
> @ash2k 
> it will be impossible to use this method and then add an indexer to the returned informer if the factory has been started already

> @danmace
> How would you safely do this today given the caller has no way of knowing whether the factory has started the informer(s)?

agree, there is no guarantee you get back an unstarted informer from this function, so adding indexers to the returned informer already has undefined behavior. should sweep to make sure we don't do that today.

Reply:
> agree, there is no guarantee you get back an unstarted informer from this function, so adding indexers to the returned informer already has undefined behavior. should sweep to make sure we don't do that today.

Wait, is there some other function to do this that I don't remember?  Today it is the only way to add indexes.  We cannot take away that power this release.

Reply:
@deads2k I actually don't see anywhere where we're adding indexes, or maybe my search is off?

Reply:
> @deads2k I actually don't see anywhere where we're adding indexes, or maybe my search is off?

Added and used downstream and the capability already shipped in client-go, at least 4.0.0, maybe earlier.

Reply:
@deads2k Hm, I don't see it as a removal of capabilities... you can still do it, and it's still unsafe. Or are you saying downstream code is relying on adding indexes to new informers created post-start?

Reply:
It is a little lucky, and we should change it, but this PR makes the initial Start() in the controller manager (which today is just starting secrets/serviceaccounts/namespaces informers) cause every subsequent Informer() created during controller setup auto-start immediately.

We should separate the bug fix from the autostart behavior as andy proposed, and see what we can do to improve adding indexes before we increase the situations where doing so would fail.

Reply:
@liggitt @ironcladlou I agree, it is already unsafe today if you don't know if the factory has been started already.

@ncdc 
> Good point. We need to decide if that's ok. I think I'd rather solve #48198 for real, instead of relying on ...Informer().AddIndexers().

Looks like this issue talks about code generation. This is only "solving" it for indexes that live either inside this repository or for generated informers. My use case is starting new informers dynamically (when a new CRD is added) with a custom index(es) added. And I don't know the types in advance so cannot pre-generate informers.
This is an event handler for a new CRD. It constructs an informer and calls `h.store.AddInformer()` which adds an index:
https://github.com/atlassian/smith/blob/c486613735b0efa18a50d64ec7f84b1b72491045/pkg/controller/controller_crd_event_handler.go#L99-L110
and
https://github.com/atlassian/smith/blob/c486613735b0efa18a50d64ec7f84b1b72491045/pkg/store/multi.go#L36-L51

After the index is added, the informer is started with `h.wg.StartWithChannel(ctx.Done(), crdInf.Run)`.

So, I don't use the factory to start informers here and anywhere really so I'm not concerned about `InformerFor()` method. But please don't remove the possibility to do what I'm doing without providing an alternative approach that will work for my use case.

p.s. I'm not using the factory because:
- I absolutely need to be able to start informers only for a single namespace. I.e. the whole program only watches things in a single namespace.
- `Start()` method starts goroutines but provides me no means to block until they finish once signaled. Hence I call `Run()` on informers myself and it is a blocking call now (after my pr has been merged) so I can track the goroutine.
p.p.s. I can imagine a use case where a new informer needs to be created with a dynamically configured index on it. Or indexes on an existing informer need to be adjusted - removed/added. Can we solve this the same (expensive?) way as dynamic event handler addition?